appengine/genetics/js/cage.js

```diff
@@ -0,0 +1,725 @@
+};
+
+/**
+ * Time to end the battle.
+ * @type {number}
+ * @private
+ */
+Genetics.Cage.endTime_ = 0;
+
+/**
+ * Time limit of game (in milliseconds).
+ * @type {number}
+ */
+Genetics.Cage.GAME_TIME_LIMIT_MSC = 5 * 60 * 1000;
+
+/**
+ * Time limit for mouse function execution in number of interpreter steps (100k
+ * ticks runs for about 3 minutes).
+ * @type {number}
+ */
+Genetics.Cage.FUNCTION_TIMEOUT_STEPS = 100000;
+
+/**
+ * The next available player ID.
+ * @type {number}
+ * @private
+ */
+Genetics.Cage.nextAvailablePlayerId_ = 0;
+
+/**
+ * The next available mouse ID.
+ * @type {number}
+ * @private
+ */
+Genetics.Cage.nextAvailableMouseId_ = 0;
+
+/**
+ * Callback function for end of game.
+ * @type {Function}
+ * @private
+ */
+Genetics.Cage.doneCallback_ = null;
+
+
+/**
+ * Stop and reset the cage.
+ */
+Genetics.Cage.reset = function() {
+  clearTimeout(Genetics.Cage.pid);
+  Genetics.Cage.EVENTS.length = 0;
+  Genetics.Cage.aliveMice_.length = 0;
+  Genetics.Cage.miceMap_ = {};
+  // Cancel all queued life simulations.
+  for (var mouseId in Genetics.Cage.lifePidsMap_) {
+    if (Genetics.Cage.lifePidsMap_.hasOwnProperty(mouseId)) {
+      clearTimeout(Genetics.Cage.lifePidsMap_[mouseId]);
+    }
+  }
+  Genetics.Cage.lifePidsMap_ = {};
+  Genetics.Cage.nextAvailableMouseId_ = 0;
+};
+
+/**
+ * Adds a player to the game.
+ * @param {string} playerName The name of the player to be added.
+ * @param {string|!Function} code Player's code, or generator.
+ */
+Genetics.Cage.addPlayer = function(playerName, code) {
+  // Assign a unique id to the player and add to the game.
+  var id = Genetics.Cage.nextAvailablePlayerId_++;
+  Genetics.Cage.players[id] = [playerName, code];
+};
+
+/**
+ * Start the Cage simulation. Players should be already added
+ * @param {Function} doneCallback Function to call when game ends.
+ */
+Genetics.Cage.start = function(doneCallback) {
+  Genetics.Cage.doneCallback_ = doneCallback;
+  // Create a mouse for each player.
+  for (var playerId in Genetics.Cage.players) {
+    if (Genetics.Cage.players.hasOwnProperty(playerId)) {
+      for (var i = 0; i < Genetics.Cage.START_MICE_PER_PLAYER; i++) {
+        var mouseSex = i % 2 == 0 ? Genetics.Mouse.Sex.MALE :
+            Genetics.Mouse.Sex.FEMALE;
+        var mouseId = Genetics.Cage.nextAvailableMouseId_++;
+        var mouse = new Genetics.Mouse(mouseId, mouseSex, null, null, playerId);
+        Genetics.Cage.born(mouse);
+        new Genetics.Cage.Event('ADD', mouse,
+            Genetics.Cage.players[playerId][0]).addToQueue();
+      }
+    }
+  }
+  Genetics.Cage.endTime_ = Date.now() + Genetics.Cage.GAME_TIME_LIMIT_MSC;
+  new Genetics.Cage.Event('START_GAME').addToQueue();
+};
+
+/**
+ * Queues life simulation for mouse.
+ * @param {Genetics.Mouse} mouse The mouse to queue the simulation for.
+ */
+Genetics.Cage.life = function(mouse) {
+  // Queue life simulation for mouse with some variation in timeout to allow for
+  // variation in order.
```

Comment:
This comment is a little cryptic. Can you elaborate more on "to allow for variation in order"?